packages/core/src/external/carequality/ihe-gateway-v2/report/report.ts

```diff
@@ -0,0 +1,301 @@
+    oid: (error.gateway as XCAGateway).homeCommunityId,
+    error:
+      error.operationOutcome?.issue?.map(issue => JSON.stringify(issue.details)).join(", ") ?? "",
+  }));
+
+  return {
+    requestId,
+    total,
+    successPercentage: total > 0 ? (successes.length / total) * 100 : 0,
+    failurePercentage: total > 0 ? (failures.length / total) * 100 : 0,
+    errorOids,
+  };
+}
+
+async function generateDocumentRetrievalReportForRequestId(
+  sequelize: Sequelize,
+  requestId: string
+): Promise<Report> {
+  const results = await getDocumentRetrievalResultsForRequestId({ sequelize, requestId });
+  const total = results.length;
+
+  const successes = results.filter(
+    (result: OutboundDocumentRetrievalResp) =>
+      (result as OutboundDocumentRetrievalResp).documentReference !== undefined
+  );
+
+  const failures = results.filter(
+    (result: OutboundDocumentRetrievalResp) =>
+      (result as OutboundDocumentRetrievalResp).documentReference === undefined
+  );
+
+  const errorOids = failures.map((error: OutboundDocumentRetrievalResp) => ({
+    oid: (error.gateway as XCAGateway).homeCommunityId,
+    error:
+      error.operationOutcome?.issue?.map(issue => JSON.stringify(issue.details)).join(", ") ?? "",
+  }));
+
+  return {
+    requestId,
+    total,
+    successPercentage: total > 0 ? (successes.length / total) * 100 : 0,
+    failurePercentage: total > 0 ? (failures.length / total) * 100 : 0,
+    errorOids,
+  };
+}
+
+function summarizeReports(reports: Report[]): ReportSummary {
+  const totalReports = reports.length;
+  const averageSuccessPercentage =
+    reports.reduce((sum, report) => sum + report.successPercentage, 0) / totalReports;
+  const averageFailurePercentage =
+    reports.reduce((sum, report) => sum + report.failurePercentage, 0) / totalReports;
+
+  const errorDetailsMap: { [key: string]: { count: number; details: string } } = {};
+  reports.forEach(report => {
+    report.errorOids.forEach(error => {
+      const key = `${error.oid}:${error.error}`;
+      const errorDetails = errorDetailsMap[key];
+      if (errorDetails) {
+        errorDetails.count += 1;
+      } else {
+        errorDetailsMap[key] = { count: 1, details: error.error ?? "" };
+      }
+    });
+  });
+
+  const simplifiedReports: SimplifiedReport[] = reports.map(report => ({
+    requestId: report.requestId,
+    total: report.total,
+    successPercentage: report.successPercentage,
+    failurePercentage: report.failurePercentage,
+  }));
+
+  return {
+    simplifiedReports,
+    averageSuccessPercentage,
+    averageFailurePercentage,
+    errorDetailsMap,
+  };
+}
+
+export async function generatePatientDiscoveryReport(sequelize: Sequelize): Promise<ReportSummary> {
+  const twelveHoursAgo = dayjs().subtract(12, "hours").toISOString();
+  const requestIds = await getRandomRequestIds({
+    sequelize,
+    tableName: "patient_discovery_result",
+    since: twelveHoursAgo,
+    limit: patientDiscoveryReportRequests,
+  });
```

Comment:
What's the thinking behind using random request IDs?

Reply:
The idea was to get a random sample of ids to look at so the statistics were representative